rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp

```diff
@@ -559,12 +588,17 @@ class RMW_Connext_Subscriber
   RMW_Connext_MessageTypeSupport * type_support;
   rmw_gid_t ros_gid;
   const bool created_topic;
-  RMW_Connext_SubscriberStatusCondition status_condition;
+  std::shared_ptr<RMW_Connext_SubscriberStatusCondition> status_condition;
   RMW_Connext_UntypedSampleSeq loan_data;
   DDS_SampleInfoSeq loan_info;
-  size_t loan_len;
+  std::atomic_uint64_t loan_len;
```

Comment:
What is the reason for making this a `std::atomic_uint64_t`? Access to `loan_len` and other loan-related fields is already protected by `std::mutex loan_mutex`.

Reply:
I think https://github.com/ros2/rmw_connextdds/blob/3e4fb9067b7bab7f7d8180ac3ed3c3054fa9933d/rmw_connextdds_common/src/common/rmw_impl.cpp#L1419-L1424 should be protected. I'll not use std::atomic_uint64 but add `std::mutex loan_mutex` for it. What do you think?

Reply:
Well, access needs to be synchronized only if applications are allowed to change the CFT settings while concurrently taking messages from the subscription (the call to `status_condition->invalidate()` should already guarantee that the subscription is not attached to any waitset, which might also end up updating the "loan state").

Before these changes, the code in `RMW_Connext_Subscriber::finalize()` assumed the operation wasn't "thread-safe" with respect to other subscription operations.

If this is not the case (i.e. an application is allowed to modify the subscription's CFT parameters while also polling the subscription for samples), then access to `loan_len` should be guarded inside `finalize()`.

By the way, I think the line `this->loan_next = this->loan_len;` can be safely removed either way (I'm not sure why it's there after reviewing the code...). Would you mind making that change?

Reply:
> 
> By the way, I think the line `this->loan_next = this->loan_len;` can be safely removed either way (I'm not sure why it's there after reviewing the code...). Would you mind making that change?

Actually, I think my intention there was to "consume all remaining samples". Still, it doesn't actually do that (or rather, there is no need to do it since the field is immediately reset by `return_messages()`...), so I would still get rid of the line.

Reply:
OK, I'll update it.